src/decompiler.cc

```diff
@@ -0,0 +1,426 @@
+        //BracketIfNeeded(stack.back());
+        // TODO: also version without () depending on precedence.
+        WrapChild(stack.back(),
+                  std::string(OpcodeToToken(ue.opcode)) + "(", ")");
+        return;
+      }
+      case ExprType::Load: {
+        auto& le = *cast<LoadExpr>(&e);
+        BracketIfNeeded(stack.back());
+        std::string suffix = "[";
+        suffix += std::to_string(le.offset);
+        suffix += "]:";
+        suffix += TypeFromLoadStore(le.opcode, string_view(".load"));
+        stack.back().v.back() += suffix;
+        // FIXME: align
+        return;
+      }
+      case ExprType::Store: {
+        auto& se = *cast<StoreExpr>(&e);
+        BracketIfNeeded(stack[stack.size() - 2]);
+        std::string suffix = "[";
+        suffix += std::to_string(se.offset);
+        suffix += "]:";
+        suffix += TypeFromLoadStore(se.opcode, string_view(".store"));
+        stack[stack.size() - 2].v.back() += suffix;
+        WrapBinary(" = ", true);
+        // FIXME: align
+        return;
+      }
+      case ExprType::If: {
+        auto ife = cast<IfExpr>(&e);
+        auto ifs = std::move(stack.back());
+        stack.pop_back();
+        DecompileExprs(ife->true_.exprs);
+        auto thenp = std::move(stack.back());
+        stack.pop_back();
+        bool multiline = ifs.v.size() > 1 || thenp.v.size() > 1;
+        size_t width = ifs.width() + thenp.width();
+        Value elsep { {}, false };
+        if (!ife->false_.empty()) {
+          DecompileExprs(ife->false_);
+          elsep = std::move(stack.back());
+          width += elsep.width();
+          stack.pop_back();
+          multiline = multiline || elsep.v.size() > 1;
+        }
+        multiline = multiline || width > target_exp_width;
+        if (multiline) {
+          auto if_start = string_view("if (");
+          ifs.v[0].insert(0, if_start.data(), if_start.size());
+          ifs.v.back() += ") {";
+          IndentValue(thenp, indent_amount, {});
+          std::move(thenp.v.begin(), thenp.v.end(), std::back_inserter(ifs.v));
+          if (!elsep.v.empty()) {
+            ifs.v.push_back("} else {");
+            IndentValue(elsep, indent_amount, {});
+            std::move(elsep.v.begin(), elsep.v.end(), std::back_inserter(ifs.v));
+          }
+          ifs.v.push_back("}");
+          stack.push_back(std::move(ifs));
+          return;
+        } else {
+          ss << "if (" << ifs.v[0] << ") { " <<  thenp.v[0] << " }";
+          if (!elsep.v.empty()) ss << " else { " << elsep.v[0] << " }";
+          break;
+        }
+      }
+      case ExprType::Loop: {
+        auto le = cast<LoopExpr>(&e);
+        DecompileExprs(le->block.exprs);
+        auto &val = stack.back();
+        IndentValue(val, indent_amount, {});
+        val.v.insert(val.v.begin(), "loop {");
+        val.v.push_back("}");
+        return;
+      }
+      default: {
+        std::string name = GetExprTypeName(e.type());
```

Comment:
Nit: can move to default case of switch and avoid unconditionally calling `GetExprTypeName`